smack-extensions/src/main/java/org/jivesoftware/smackx/usertune/provider/UserTuneProvider.java

```
@@ -0,0 +1,115 @@
+                    continue outerloop;
+                }
+                try {
+                    while (true) {
+                        switch (name) {
+                        case "artist":
+                            builder.setArtist(parser.nextText());
+                            break;
+                        case "length":
+                            builder.setLength(Integer.parseInt(parser.nextText()));
+                            break;
+                        case "rating":
+                            builder.setRating(Integer.parseInt(parser.nextText()));
+                            break;
+                        case "source":
+                            builder.setSource(parser.nextText());
+                            break;
+                        case "title":
+                            builder.setTitle(parser.nextText());
+                            break;
+                        case "track":
+                            builder.setTrack(Integer.parseInt(parser.nextText()));
+                            break;
+                        case "uri":
+                            URI uri = new URI(parser.nextText());
+                            builder.setUri(uri);
+                            break;
+                        case "tune":
+                            if (UserTuneElement.ELEMENT.equals(name) && (tag == Event.END_ELEMENT)) {
+                                return builder.build();
+                            }
+                        }
+                        tag = parser.next();
+                        name = parser.getName();
+                        if (isNameInvalid(name)) {
+                            throw new XmlPullParserException("Invalid elements inside <tune/>");
+                        }
+                    }
+                } catch (IllegalArgumentException e) {
+                    throw new XmlPullParserException("Unknown tune value: " + name + " encountered.");
```

You don't have to catch this exception, simply let it throw.